pkg/test/framework/components/echo/check/checkers.go

```diff
@@ -363,10 +365,17 @@ func URL(expected string) echo.Checker {
 
 // ReachedTargetClusters is similar to ReachedClusters, except that the set of expected clusters is
 // retrieved from the Target of the request.
-func ReachedTargetClusters(allClusters cluster.Clusters) echo.Checker {
+func ReachedTargetClusters(t framework.TestContext) echo.Checker {
+	dnsCaptureEnabled := istio.GetOrFail(t, t).MeshConfigOrFail(t).DefaultConfig.ProxyMetadata["ISTIO_META_DNS_CAPTURE"] == "true"
 	return func(result echo.CallResult, err error) error {
+		if !dnsCaptureEnabled && result.Opts.To.Config().IsHeadless() {
+			// Special case: requests to headless services will only go cross-cluster when
+			// DNS capture is enabled.
+			return checkReachedSourceClusterOnly(result, t.AllClusters())
+		}
+
 		expectedByNetwork := result.Opts.To.Clusters().ByNetwork()
-		return checkReachedClusters(result, allClusters, expectedByNetwork)
+		return checkReachedClusters(result, t.AllClusters(), expectedByNetwork)
```

Comment:
I don't think we deploy any workload in external control plane cluster, `t.Clusters()` would be good.